webapp/src/components/harbour/QuotaTokenBalance.tsx

```diff
@@ -0,0 +1,140 @@
+import { zodResolver } from "@hookform/resolvers/zod";
+import { ethers, type JsonRpcApiProvider, toBeHex } from "ethers";
+import { useForm } from "react-hook-form";
+import { z } from "zod";
+import { Box, BoxTitle } from "@/components/Groups";
+import { useBrowserProvider } from "@/hooks/useBrowserProvider";
+import { useQuotaTokenStats } from "@/hooks/useQuotaStats";
+import { switchToChain } from "@/lib/chains";
+import { getShortAddress } from "@/lib/encoding";
+import { type ERC20TokenDetails, encodeERC20Approval } from "@/lib/erc20";
+import { getHarbourChainId, harbourAt } from "@/lib/harbour";
+import { positiveAmountSchema } from "@/lib/validators";
+import { FormItem, SubmitItem } from "../Forms";
+
+const createTokenTopUpFormSchema = () =>
+	z.object({
+		amount: positiveAmountSchema,
+	});
+
+type TokenTopUpFormData = z.infer<
+	ReturnType<typeof createTokenTopUpFormSchema>
+>;
+
+function buildAmountLabel(
+	tokenInfo: Partial<ERC20TokenDetails> | undefined,
+): string {
+	if (tokenInfo?.balance === undefined) return "Amount";
+	return `Amount (Balance: ${ethers.formatUnits(tokenInfo.balance ?? 0, tokenInfo.decimals)})`;
+}
+
+function QuotaTokenBalance({
+	signerAddress,
+	harbourProvider,
+	harbourAddress,
+	className,
+}: {
+	signerAddress: string | undefined;
+	harbourProvider: JsonRpcApiProvider | null;
+	harbourAddress: string | undefined;
+	className?: string;
+}) {
+	const provider = useBrowserProvider();
+	const { quotaTokenStats, isLoading } = useQuotaTokenStats(
+		harbourProvider,
+		signerAddress,
+		harbourAddress,
+	);
+
+	const {
+		register,
+		handleSubmit,
```

Comment:
I was using `useTransition` to keep track of pending async work on button submit - the `zod` + `react-hook-form` seems super useful.

Generally speaking (as a non-React developer), I find discovery of these tools quite hard 😅. It might make sense for us to maintain in a repo somewhere a meta NPM package with general React tools that we use across our projects. Might be more effort than its worth, and maybe a quick "sharing" session might be enough.